File: go/sorting/selection_sort.go
package sorting

type SelectionSort struct {
}

func (s SelectionSort) Sort(arr []int, sortType SortType) []int {
	length := len(arr)

	for i := 0; i < length; i++ {
		minIndex := i
		for j := i + 1; j < length; j++ {
			if Compare(arr[minIndex], arr[j], sortType) {
				minIndex = j
			}
		}
		arr[i], arr[minIndex] = arr[minIndex], arr[i]
	}

	return arr
}

File: scala/src/main/scala/sorting/SelectionSort.scala
package sorting

class SelectionSort() extends Sort {
  def sort(arr: Array[Int], sortType: SortType): Array[Int] = {
    val length: Int = arr.length

    for (i <- Range(0, length - 1, 1)) {
      var minIndex = i
      for (j <- Range(i + 1, length, 1))
        if (compare(arr(minIndex), arr(j), sortType)) {
          minIndex = j
        }
      swap(arr, i, minIndex)
    }
    arr
  }
}
